### PR TITLE
ENH: Add Docker authentication to increase pull rate limit

### LINKED ISSUE #2311: [Action Required] How to avoid Docker Hub rate limits for your CircleCI builds
Just forwarding an email from CircleCI:

> Hi there,
> 
> On November 1st, Docker Hub will begin limiting anonymous image pulls. We want to make sure you know how you might be impacted and what you can do to avoid interruptions to your workflow.
> 
> Adding Docker authentication to your pipeline config is the easiest way to avoid any service disruptions. If you use the Docker executor or pull Docker images when using the machine executor on CircleCI, we encourage you to authenticate. Because the anonymous API rate limits are based on IP addresses, they will impact CircleCI cloud customers. Authenticated users get higher per-user rate limits, regardless of IP.
> 
> We are currently working on a partnership with Docker to minimize the impact of this change for our users and will share more details as we get them.
> 
> For more information or to leave a question for us, please head over to Discuss or contact support.
> 
> Thanks and happy building,
> 
> \- The CircleCI team

We can either start using GH Packages and pull from their registry (probably we'll enjoy more bandwidth even now) or add the login.

### PR DESCRIPTION
**Pull request by mgxd**
Partially? closes #2311 

https://circleci.com/docs/2.0/private-images/

### CODE REVIEW

**Comment by effigies** [comment]
Looks like this is going to be a problem with PRs, since we would be exposing credentials to any fork.

**Comment by mgxd** [comment]
Good point - this has been [brought up](https://discuss.circleci.com/t/updated-authenticate-with-docker-to-avoid-impact-of-nov-1st-rate-limits/37567/49) in the circle discussion forum, but no workaround for the `machine` executor has been proposed yet.

**Comment by effigies** [comment]
We can still wrap everything into a `if [[ -n "$DOCKER_PASS" ]]` block. That will boost speeds on the main repository's builds, and then PRs can use the caches, and only rarely hit Docker Hub.

**Comment by mgxd** [comment]
Yup, though my worry is PR builds could hit the rate limit (I believe 100 pulls/day). I don't expect us to normally hit that cap, but something to monitor.

**Comment by effigies** [comment]
Is that a serious concern, though? Most of our pulls are from the local registry, which is loaded from a cache. As long as `master` is building those caches regularly, there's little danger of even 1 pull per day from PRs.

**Comment by mgxd** [comment]
I was under the impression every "Set-up Docker registry" step in our build is pulling the `registry` image - i.e. https://app.circleci.com/pipelines/github/nipreps/fmriprep/33/workflows/71306b88-b53e-4a3e-bef8-070d71ad5700/jobs/222/parallel-runs/0/steps/0-107

**Comment by effigies** [comment]
Oh, good call. Maybe we can tarball that image.

File: .circleci/config.yml [hunk 1]
```diff
@@ -26,9 +25,25 @@ jobs:
             - build-v4-
           paths:
             - /tmp/docker
+            - /tmp/images
```
Comment (effigies) [inline .circleci/config.yml:28]: Is it possible we'll want to clear the `registry:2` cache independent of the build cache?

**Review by effigies — APPROVED** [review]
LGTM. I think we could separate out the `registry:2` cache, but I don't think it's critical to do so.

File: .circleci/config.yml (hunk 1 above)
Comment (mgxd) [reply .circleci/config.yml:28]: I think we'll update that image so infrequently that it's not really worth it to make a new cache 🤷‍♂️

**Comment by effigies** [comment]
Let's do it. Can always fix, and it'd be good to confirm that PRs are avoiding hitting docker hub.

**Comment by effigies** [comment]
... can't merge from my phone. Feel free to merge.

**Comment by oesteban** [comment]
I have merged because I think this is good in the long term. However it seems CircleCI has partnered with Docker Hub to avoid exactly this problem - https://discuss.circleci.com/t/updated-authenticate-with-docker-to-avoid-impact-of-nov-1st-rate-limits/37567?mkt_tok=eyJpIjoiTVRsaU1HWmtNRE16TkRZdyIsInQiOiJWcVBBMFJZdlR3TFhSVmxYejJtbDJKa2dqY3ZoY1k0TVl1eUs4d3Vpc2dqOG1ZSTNibHVGd1pTMXFJeGhlTkdzcnJQRWtrcWIwZjc0N1BEVUlPSFNNXC9qdVFxOFpcLzZhUGFHS0x1RHNLVXNGak1id25WaUVcL3FiYktIWlpqcTgrOHNGdDRJaldKdlUrQWNIVmVJWVUxdUE9PSJ9

I just got the email, literally.

**Comment by mgxd** [comment]
@oesteban I think the limit still applies to `machine` executors, which is what we're using (hopefully to take advantage of docker layer caching down the line).

https://discuss.circleci.com/t/updated-authenticate-with-docker-to-avoid-impact-of-nov-1st-rate-limits/37567/52
https://discuss.circleci.com/t/updated-authenticate-with-docker-to-avoid-impact-of-nov-1st-rate-limits/37567/58?u=mgxd

**Pull request merged**